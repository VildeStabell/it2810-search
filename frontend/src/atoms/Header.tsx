import React, { useState } from "react";
import store from "../redux/store";
import "../styles/Header.css";
import RegisterUserForm from "./RegisterUserForm";

/**
 * Header is a component containing every item in the header. What buttons
 * you see will depend on whethter or not you are logged in.
 * @var isLoggedIn keeps track of if the user is logged in or not.
 */

// store.dispatch(changeLoginStatus(true));

const Header = () => {
  const [isloggedIn, setIsLoggedIn] = useState<boolean>(
    store.getState().loginStatusReducer.loginStatus
  );

  const [displayRegisterForm, updateDisplayRegisterForm] = useState<boolean>(
    false
  );

  store.subscribe(() =>
    setIsLoggedIn(store.getState().loginStatusReducer.loginStatus)
  );

  const toggleRegisterForm = () => {
    updateDisplayRegisterForm(!displayRegisterForm);
  };

  return (
    <header>
      <h3 id="title">[SETT INN EN GOD TITTEL HER]</h3>
      <div className="header-right">
        {isloggedIn ? (
          <button id="profile-button">PROFILE</button>
        ) : (
          <button id="register-button" onClick={toggleRegisterForm}>
            REGISTER
          </button>
        )}
        {isloggedIn ? (
          <button className="login-buttons">SIGN OUT</button>
        ) : (
          <button className="login-buttons">SIGN IN</button>
        )}
        {displayRegisterForm ? (
          <RegisterUserForm toggleRegisterForm={toggleRegisterForm} />
        ) : null}
      </div>
    </header>
  );
};

export default Header;
