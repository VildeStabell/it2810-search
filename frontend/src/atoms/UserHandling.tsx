import React from "react";

const UserHandling = () => {
  const displayLoginForm = () => {};

  const displayRegisterForm = () => {};
  return (
    <>
      <button onClick={displayLoginForm}>Sign in to existing user</button>
      <button onClick={displayRegisterForm}>Create new user</button>
    </>
  );
};

export default UserHandling;
