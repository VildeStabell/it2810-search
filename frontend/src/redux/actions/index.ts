export const UPDATE_DETAILED_BOOK_ID = "UPDATE_DETAILED_BOOK_ID";
export const CHANGE_LOGIN_STATUS = "CHANGE_LOGIN_STATUS";
export const UPDATE_PHONE_PAGE = "UPDATE_PHONE_PAGE";
export const UPDATE_BOOK_PAGE = "UPDATE_BOOK_PAGE";
export const CHANGE_SEARCH = "CHANGE_SEARCH";

export const changeDetailedBook = (id: string) => ({
  type: UPDATE_DETAILED_BOOK_ID,
  id,
});

export const changeLoginStatus = (loginStatus: boolean) => ({
  type: CHANGE_LOGIN_STATUS,
  loginStatus,
});

export const changePhonePage = (newPage: string) => ({
  type: UPDATE_PHONE_PAGE,
  newPage,
});

export const changeBookPage = (newPage: number) => ({
  type: UPDATE_BOOK_PAGE,
  newPage,
});

export const changeSearch = (searchString: string) => ({
  type: CHANGE_SEARCH,
  searchString,
});
