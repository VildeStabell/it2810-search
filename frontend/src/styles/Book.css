#book-container {
  margin-right: 21vw;
  display: flex;
  flex-wrap: wrap;
  background-color: #ffffff;
  justify-content: space-evenly;
}

.single-book {
  display: inline-block;
  margin: 5px;
  padding: 10px;
  text-align: center;
  border: 1px solid #eeeeee;
}

.single-book img {
  width: 150px;
  height: 200px;
  object-fit: cover;
  object-position: 45%;
}

.single-book img:hover {
  cursor: pointer;
}

.single-book .title {
  font-style: italic;
}

.single-book .title:hover {
  cursor: pointer;
}

.single-book .author {
  margin: 0;
}

@media only screen and (max-width: 850px) {
  .single-book img {
    width: 100px;
    height: 150px;
  }

  .single-book {
    width: 35%;
  }

  #book-container {
    margin: 0;
  }
}
