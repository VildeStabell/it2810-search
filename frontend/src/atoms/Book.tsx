import React from "react";
import { useDispatch } from "react-redux";
import { changeDetailedBook, changePhonePage } from "../redux/actions";
import "../styles/Book.css";

export interface BookProps {
  id: string;
  title: string;
  author: string;
  cover: string;
}

/**
 * Book is a representation of a book, with its title,
 * author and cover image.
 * @param props
 */
export const Book = (props: BookProps) => {
  const dispatch = useDispatch();

  const handleBookClick = (id: string) => {
    updateDetailedBook(id);
    dispatch(changePhonePage("book"));
  };

  const updateDetailedBook = (id: string) => {
    dispatch(changeDetailedBook(id));
  };

  return (
    <div
      id={String(props.id)}
      data-testid="single-book"
      className="single-book"
      onClick={() => handleBookClick(props.id)}
    >
      <img src={props.cover} alt="`SOME book cover" />
      <p className="title" data-testid="book-title">{props.title}</p>
      <p className="author" data-testid="book-author">by {props.author}</p>
    </div>
  );
};
