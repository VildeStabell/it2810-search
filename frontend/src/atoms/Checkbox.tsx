import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { addFilter, removeFilter } from "../redux/actions";

interface CheckboxProps {
  text: string;
  id: string;
}

/**
 * Checkbox is a component to create a checkbox with text beside it,
 * This code is inspired by https://www.w3schools.com/howto/tryit.asp?filename=tryhow_css_custom_checkbox
 * @param text is the text right of the checkbox.
 * @param id is the id of the checkbox.
 * @var checked is a hook to decide if a checkbox is checked.
 * @var filters is a list containing the current filers being used.
 */

const Checkbox = (props: CheckboxProps) => {
  const [checked, changeChecked] = useState(false);
  const dispatch = useDispatch();
  const filters: any = useSelector((state: any) => state.filter.filters);

  const toggleCheck = () => {
    changeChecked(!checked);
    if (filters.includes(props.text)) {
      dispatch(removeFilter(props.text));
    } else {
      dispatch(addFilter(props.text));
    }
  };

  return (
    <span id={props.id} className="container" onClick={toggleCheck}>
      {props.text}
      <input type="checkbox" checked={checked} />
      <span className="checkmark"></span>
    </span>
  );
};

export default Checkbox;
