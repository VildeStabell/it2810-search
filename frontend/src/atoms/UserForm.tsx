import React, { useState } from "react";
import { MdClose } from "react-icons/md";
import { useDispatch, useSelector } from "react-redux";
import { changeLoginStatus } from "../redux/actions";
import "../styles/UserForm.css";

interface UserFormProps {
  isLoginForm: boolean;
  toggleForm: () => void;
}

const API_URL = "http://localhost:3002/auth/";

const UserForm = (props: UserFormProps) => {
  const [username, setUsername] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const [passwordConfirmation, setPasswordConfirmation] = useState<string>("");
  const [correctFormInput, setCorrectFormInput] = useState<string>("");

  const phonePage: any = useSelector((state: any) => state.phonePage.phonePage);
  const dispatch = useDispatch();

  const validateInput = (e: any) => {
    e.preventDefault();

    if (username && password) {
      if (props.isLoginForm) {
        registerUser();
      } else {
        if (passwordConfirmation && password === passwordConfirmation) {
          registerUser();
        } else
          setCorrectFormInput("Password and Confirmed password do not match!");
      }
    } else setCorrectFormInput("Please fill out all fields.");
  };

  const registerUser = () => {
    fetch(props.isLoginForm ? API_URL + "login" : API_URL + "register", {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        username: username,
        password: password,
      }),
    })
      .then((res) => {
        if (res.ok) {
          dispatch(changeLoginStatus(true));
          props.toggleForm();

          return res.json();
        }
      })
      .catch((err) => console.log(err));
  };

  const handleInputChange = (e: any) => {
    if (e.target.name === "username") {
      setUsername(e.target.value);
    } else if (e.target.name === "password") {
      setPassword(e.target.value);
    } else if (e.target.name === "confirm-password") {
      setPasswordConfirmation(e.target.value);
    }
    setCorrectFormInput("");
  };

  return (
    <div className={phonePage}>
      <button
        id="popup-close-button"
        className="red-button"
        onClick={props.toggleForm}
      >
        <MdClose size="20px" />
      </button>
      <h3>{props.isLoginForm ? "Sign in" : "Register user"}</h3>
      <form>
        <label>Username</label>
        <input
          id="username-input"
          name="username"
          value={username}
          type="text"
          placeholder="Type username"
          onChange={handleInputChange}
          required
        />
        <br />
        <label>Password</label>
        <input
          id="password-input"
          name="password"
          value={password}
          type="password"
          placeholder="Type password"
          onChange={handleInputChange}
          required
        />
        <br />
        {props.isLoginForm ? null : (
          <>
            <label>Confirm password</label>
            <input
              id="confirm-password-input"
              name="confirm-password"
              value={passwordConfirmation}
              type="password"
              placeholder="Type password again"
              onChange={handleInputChange}
              required
            />
          </>
        )}
        <br />
        <b>{correctFormInput}</b>
        <br />
        <input type="submit" className="red-button" onClick={validateInput} />
      </form>
    </div>
  );
};

export default UserForm;
