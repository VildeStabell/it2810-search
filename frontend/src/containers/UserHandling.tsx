import React, { useState } from "react";
import UserForm from "../atoms/UserForm";

const UserHandling = () => {
  const [displayRegisterForm, changeDisplayRegisterForm] = useState<boolean>(
    false
  );

  return (
    <div id="userHandling">
      {displayRegisterForm ? (
        <>
          <UserForm isLoginForm={false} />
          Already got a user? Sign in below!
          <button
            id="userHandlingLogin"
            onClick={() => changeDisplayRegisterForm(false)}
            className="red-button"
          >
            Sign in to existing user
          </button>
        </>
      ) : (
        <>
          <UserForm isLoginForm={true} />
          Don't have a user? Create one!
          <button
            id="userHandlingRegister"
            onClick={() => changeDisplayRegisterForm(true)}
            className="red-button"
          >
            Create new user
          </button>
        </>
      )}
    </div>
  );
};

export default UserHandling;
