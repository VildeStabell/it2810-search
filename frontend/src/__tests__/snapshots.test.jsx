import React from "react";
import * as reactModule from "react";
import * as reduxModule from "react-redux";
import { cleanup } from "@testing-library/react";
import ShallowRenderer from "react-shallow-renderer";
import { Book } from "../atoms/Book";
import UserHandling from "../containers/UserHandling";
import UserForm from "../atoms/UserForm";
import { changeLoginStatus } from "../redux/actions";
import loginStatusReducer, {
  initialState,
} from "../redux/reducers/loginReducer";

afterEach(cleanup);

/**
 * Snapshot tests for Login. Are testing the reducer changeLoginStatus and the components UserHandling and UserForm.
 * The tests are inspired by:
 * Handle hooks: https://medium.com/@codegagan/react-functional-component-testing-by-mocking-hooks-55003c7ba0d7
 * Reducers: https://medium.com/@bryanfillmer/reducer-snapshots-for-rapid-development-2ede745b8c40
 */

describe("Snapshot test of reducer", () => {
  it("should give loginStatus true", () => {
    const actual = loginStatusReducer(initialState, changeLoginStatus(true));
    expect(actual).toMatchSnapshot();
  });
});

describe("Snapshot test of UserHandling when displayRegisterForm=true", () => {
  it("should render a userForm containing the text 'Already got a user? Sign in below!', and a button with the text 'Sign in to existing user'", () => {
    const registerFormValue = true;
    reactModule.useState = jest.fn((initialRegisterFormValue) => [
      registerFormValue,
      () => {},
    ]);
    const renderer = new ShallowRenderer();
    const container = renderer.render(<UserHandling />);
    expect(container).toMatchSnapshot();
  });
});

describe("Snapshot test of UserHandling when displayRegisterForm=false", () => {
  it("should render a userForm containing the text 'Don't have a user? Create one!', and a button with the text 'Create new user'", () => {
    const registerFormValue = false;
    reactModule.useState = jest.fn((initialRegisterFormValue) => [
      registerFormValue,
      () => {},
    ]);
    const renderer = new ShallowRenderer();
    const container = renderer.render(<UserHandling />);
    expect(container).toMatchSnapshot();
  });
});

describe("Snapshot test of UserForm when props.isLoginForm=true", () => {
  let container;
  let setUsername;
  let setPassword;
  let setPasswordConfirmation;
  let setCorrectFormInput;

  beforeEach(() => {
    setUsername = jest.fn((x) => {});
    setPassword = jest.fn((x) => {});
    setPasswordConfirmation = jest.fn((x) => {});
    setCorrectFormInput = jest.fn((x) => {});

    reactModule.useState = jest
      .fn()
      .mockImplementationOnce((x) => [x, setUsername])
      .mockImplementationOnce((x) => [x, setPassword])
      .mockImplementationOnce((x) => [x, setPasswordConfirmation])
      .mockImplementationOnce((x) => [x, setCorrectFormInput]);
  });

  it("should render with an <h3> with text 'Sign in', and no field for confirming password", () => {
    reduxModule.useSelector = jest.fn(() => {});
    reduxModule.useDispatch = jest.fn(() => {});
    const props = {
      isLoginForm: true,
      toggleForm: () => {},
    };
    const renderer = new ShallowRenderer();
    container = renderer.render(<UserForm {...props} />);
    expect(container).toMatchSnapshot();
  });
});

describe("Snapshot test of UserForm when props.isLoginForm=false", () => {
  let container;
  let setUsername;
  let setPassword;
  let setPasswordConfirmation;
  let setCorrectFormInput;

  beforeEach(() => {
    setUsername = jest.fn((x) => {});
    setPassword = jest.fn((x) => {});
    setPasswordConfirmation = jest.fn((x) => {});
    setCorrectFormInput = jest.fn((x) => {});

    reactModule.useState = jest
      .fn()
      .mockImplementationOnce((x) => [x, setUsername])
      .mockImplementationOnce((x) => [x, setPassword])
      .mockImplementationOnce((x) => [x, setPasswordConfirmation])
      .mockImplementationOnce((x) => [x, setCorrectFormInput]);
  });

  it("should render with an <h3> with text 'Register user', and a field for confirming password", () => {
    reduxModule.useSelector = jest.fn(() => {});
    reduxModule.useDispatch = jest.fn(() => {});
    const props = {
      isLoginForm: false,
      toggleForm: () => {},
    };
    const renderer = new ShallowRenderer();
    container = renderer.render(<UserForm {...props} />);
    expect(container).toMatchSnapshot();
  });
});

// Snapshot test for Book:

describe("Snapshot test for book", () => {
  it("should render a book with title 'Harry Potter and the Sorcerers stone'", () => {
    const props = {
      id: "f32i30htrngf",
      title: "Harry Potter and the Sorcerers stone",
      author: "J. K. Rowling",
      cover: "urlforimage.com",
    };
    const dispatchValue = () => {};
    reduxModule.useDispatch = jest.fn(dispatchValue);
    const renderer = new ShallowRenderer();
    const container = renderer.render(<Book {...props} />);
    expect(container).toMatchSnapshot();
  });
});
