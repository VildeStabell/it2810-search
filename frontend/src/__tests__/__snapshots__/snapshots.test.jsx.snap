// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for book should render a book with title 'Harry Potter and the Sorcerers stone' 1`] = `
<div
  className="single-book"
  id="f32i30htrngf"
  onClick={[Function]}
>
  <img
    alt="\`SOME book cover"
    src="urlforimage.com"
  />
  <p
    className="title"
  >
    Harry Potter and the Sorcerers stone
  </p>
  <p
    className="author"
  >
    by 
    J. K. Rowling
  </p>
</div>
`;

exports[`Snapshot test of UserForm when props.isLoginForm=false should render with an <h3> with text 'Register user', and a field for confirming password 1`] = `
<div>
  <button
    className="red-button"
    id="popup-close-button"
    onClick={[Function]}
  >
    <MdClose
      size="20px"
    />
  </button>
  <h3>
    Register user
  </h3>
  <form>
    <label>
      Username
    </label>
    <input
      id="username-input"
      name="username"
      onChange={[Function]}
      placeholder="Type username"
      required={true}
      type="text"
      value=""
    />
    <br />
    <label>
      Password
    </label>
    <input
      id="password-input"
      name="password"
      onChange={[Function]}
      placeholder="Type password"
      required={true}
      type="password"
      value=""
    />
    <br />
    <React.Fragment>
      <label>
        Confirm password
      </label>
      <input
        id="confirm-password-input"
        name="confirm-password"
        onChange={[Function]}
        placeholder="Type password again"
        required={true}
        type="password"
        value=""
      />
    </React.Fragment>
    <br />
    <b>
      
    </b>
    <br />
    <input
      className="red-button"
      onClick={[Function]}
      type="submit"
    />
  </form>
</div>
`;

exports[`Snapshot test of UserForm when props.isLoginForm=true should render with an <h3> with text 'Sign in', and no field for confirming password 1`] = `
<div>
  <button
    className="red-button"
    id="popup-close-button"
    onClick={[Function]}
  >
    <MdClose
      size="20px"
    />
  </button>
  <h3>
    Sign in
  </h3>
  <form>
    <label>
      Username
    </label>
    <input
      id="username-input"
      name="username"
      onChange={[Function]}
      placeholder="Type username"
      required={true}
      type="text"
      value=""
    />
    <br />
    <label>
      Password
    </label>
    <input
      id="password-input"
      name="password"
      onChange={[Function]}
      placeholder="Type password"
      required={true}
      type="password"
      value=""
    />
    <br />
    <br />
    <b>
      
    </b>
    <br />
    <input
      className="red-button"
      onClick={[Function]}
      type="submit"
    />
  </form>
</div>
`;

exports[`Snapshot test of UserHandling when displayRegisterForm=false should render a userForm containing the text 'Don't have a user? Create one!', and a button with the text 'Create new user' 1`] = `
<div
  id="userHandling"
>
  <React.Fragment>
    <UserForm
      isLoginForm={true}
      toggleForm={[Function]}
    />
    Don't have a user? Create one!
    <button
      className="red-button"
      id="userHandlingRegister"
      onClick={[Function]}
    >
      Create new user
    </button>
  </React.Fragment>
</div>
`;

exports[`Snapshot test of UserHandling when displayRegisterForm=true should render a userForm containing the text 'Already got a user? Sign in below!', and a button with the text 'Sign in to existing user' 1`] = `
<div
  id="userHandling"
>
  <React.Fragment>
    <UserForm
      isLoginForm={false}
      toggleForm={[Function]}
    />
    Already got a user? Sign in below!
    <button
      className="red-button"
      id="userHandlingLogin"
      onClick={[Function]}
    >
      Sign in to existing user
    </button>
  </React.Fragment>
</div>
`;

exports[`Snapshot test of reducer should give loginStatus true 1`] = `
Object {
  "loginStatus": true,
}
`;
